File: src/main/java/com/intellij/tapestry/intellij/view/nodes/RootNode.java
package com.intellij.tapestry.intellij.view.nodes;

import com.intellij.openapi.module.Module;
import com.intellij.openapi.project.Project;
import com.intellij.tapestry.intellij.util.TapestryUtils;
import com.intellij.ui.treeStructure.SimpleNode;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

/**
 * Tapestry view root node.
 */
public class RootNode extends SimpleNode {

    private static final String ID = "ROOT";
    protected Object _myElement;

    public RootNode(@NotNull final Project project) {
        super(project);

        _myElement = ID;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public Object getElement() {
        return _myElement;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public SimpleNode @NotNull [] getChildren() {
        final List<AbstractModuleNode> newNodes = new ArrayList<>();
        final Module[] allTapestryModules = TapestryUtils.getAllTapestryModules(myProject);

        for (final Module module : allTapestryModules)
            newNodes.add(new ModuleNode(module));


        return newNodes.<AbstractModuleNode>toArray(new AbstractModuleNode[0]);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public Object @NotNull [] getEqualityObjects() {
        return new Object[]{_myElement};
    }
}

File: src/main/kotlin/com/github/rar91279/plugin/tapestry/intellij/view/nodes/RootNode.kt
package com.github.rar91279.plugin.tapestry.intellij.view.nodes

import com.intellij.openapi.project.Project
import com.github.rar91279.plugin.tapestry.intellij.util.TapestryUtils
import com.intellij.ui.treeStructure.SimpleNode

/**
 * Tapestry view root node.
 */
class RootNode(project: Project) : SimpleNode(project) {

    // Used only for getEqualityObjects() (node-identity diffing across tree refreshes). The root has
    // no backing domain object; getElement() is intentionally left at the SimpleNode default (`this`)
    // — see TapestryNode.getValue() for why overriding it here would break the async tree model.
    private val id = "ROOT"

    override fun getChildren(): Array<SimpleNode> =
        TapestryUtils.getAllTapestryModules(myProject).map { ModuleNode(it) as SimpleNode }.toTypedArray()

    override fun getEqualityObjects(): Array<Any> = arrayOf(id)
}
